feat: JWT 인증 구현

kimyu0218 opened the pull request:
resolved #5

## 작업 개요
<!-- 작업에 대한 설명을 간단하게 작성해주세요. -->
- JWT 인증 구현

## 작업 사항
<!-- 작업에 대한 설명을 코드와 관련하여 남겨주세요. -->
- JWT 토큰 생성 및 유효성 검사를 위한 유틸 구현
- JWT 필터 구현
- refresh 토큰을 위한 저장소 구현

## 고민한 점들(필수 X)
<!-- 작업을 진행하면서 고민했던 점들을 추가해주세요 -->
- 계층, 도메인 디렉토리 구조만 이용해보다가 처음 `presentation - application - ~` 방식을 사용해봐서 나중에 잘못된 부분들 옮기겠습니다 😭 
- refresh 토큰을 한 달 기준으로 잡았더니 `long` 범위를 초과해서 14일로 해놓았습니다..! 오늘 논의해보면 좋을 것 같아요 :)

## 스크린샷(필수 X)
<!-- 작업을 파악하는 데 도움이 되는 스크린샷을 추가해주세요 -->
여기에 작성하세요

kimyu0218 committed:
feat: JWT 유틸리티 클래스 정의

- generateTokens: access 토큰, refresh 토큰 생성
- validateToken: 토큰 유효성 검사
- getSubject: 토큰에서 subject 추출
- isExpired: 토큰 만료 여부 확인

kimyu0218 committed:
feat: JWT 클래스 정의

access 토큰, refresh 토큰, refresh 토큰 만료 기한 필드 정의

kimyu0218 committed:
feat: JWT 필터 정의

request 헤더에서 JWT 토큰을 추출하고 유효성 검사
- 로그인은 인증 받지 않은 사용자가 이용할 수 있으므로 whitelist에
  추가하여 필터링 예외 처리
- 토큰 유효성 검사 후 subject 정보를 request의 uuid 속성에 저장하여
  controller 단에서 이용할 수 있도록 구현
- 토큰 유효성 검사가 실패한 경우 401 반환

kimyu0218 committed:
feat: JWT 필터 구성 추가

"/api/*"경로에 대해 JwtFilter 적용

kimyu0218 committed:
feat: Redis 기반 refresh token 저장소 구현

- save: refresh 토큰 저장
- find: refresh 토큰 조회

kimyu0218 committed:
chore: 환경 변수 업데이트

- Redis 연결 설정 추가
- JWT 설정 추가

kimyu0218 commented on build.gradle:
```diff
@@ -41,6 +42,7 @@ dependencies {
 
 	// jwt
 	implementation 'io.jsonwebtoken:jjwt:0.9.1'
+	implementation 'javax.xml.bind:jaxb-api:2.3.1'
```
여기 충돌날 것 같아서 나중에 머지할 때 이 부분 제외하면 될 것 같아요!

songyi00 commented on src/main/java/com/nexters/goalpanzi/config/jwt/Jwt.java:
```diff
@@ -0,0 +1,18 @@
+public class Jwt {
+    private final String accessToken;
+    private final String refreshToken;
+    private final long refreshExpiresIn;
```
null 체크가 가능하도록 primitive 타입보다는 wrapper 클래스 타입을 쓰면 어떨까 ?!

songyi00 commented on src/main/java/com/nexters/goalpanzi/config/jwt/JwtUtil.java:
```diff
@@ -0,0 +1,72 @@
+
+@Slf4j
+@Component
+public class JwtUtil {
```
Util 네이밍보다는 JwtProvider나 JwtManager와 같은 네이밍이 조금 더 객체지향적일 것 같아!
토큰을 생성하고 검증하는 책임이 있는 클래스니까 !!

songyi00 commented on src/main/java/com/nexters/goalpanzi/domain/refreshToken/RefreshTokenRepository.java:
```diff
@@ -0,0 +1,31 @@
+import java.time.Duration;
+
+@Repository
+public class RefreshTokenRepository {
```
요거 속해있는 domain 패키지 이름 auth는 어떨까 ?!
auth 패키지에 관련된 도메인들 함께 넣어도 좋을 것 같아ㅎㅎ

songyi00 commented on src/main/java/com/nexters/goalpanzi/config/jwt/JwtFilter.java:
```diff
@@ -0,0 +1,85 @@
+import java.io.IOException;
+import java.util.List;
+
+public class JwtFilter extends GenericFilterBean {
```
사용자 요청 시 딱 한번만 필터를 타는 것을 직관적으로 나타낼 수 있도록
`GenericFilterBean` 보다는 `OncePerRequestFilter`를 사용해보는건 어떨까 ??
[참고 링크](https://minkukjo.github.io/framework/2020/12/18/Spring-142/)

songyi00 reviewed (approved):
고생했어 :+1::+1:
전체적인 구조는 PR merge 하고 조금 더 추상화시켜보자!

kimyu0218 committed:
style: 코드 포맷터 변경

kimyu0218 committed:
refactor: record로 변경

kimyu0218 committed:
feat: RedisRepository 인터페이스 정의

kimyu0218 committed:
rename: JwtUtil -> JwtManager로 변경

- JWT의 상태를 관리하고 조정하므로 manager로 변경
- claims 파싱 과정에서 발생할 수 있는 에러 처리 추가
- 만료된 토큰에 대해서는 claim 정보를 조회할 수 있도록 부분 허용 (추후
  만료된 access token에서 subject 추출 후 refresh token 검증에 이용)

kimyu0218 committed:
refactor: 테스트 케이스 추가 및 함수명 변경

- 테스트 함수 한글로 통일
- claims 파싱 과정에서 발생할 수 있는 예외에 대한 테스트 추가

kimyu0218 committed:
refactor: validation 추가

kimyu0218 committed:
refactor: assertion 함수 변경

kimyu0218 committed:
feat: Fitler 상위 인터페이스 변경 및 에러 처리 추가

- GenericFilterBean -> OncePerRequestFilter
- RequiredArgsConstructor 활용
- 만료되지 않은 유효하지 않은 토큰에 대해 400 던지도록 수정

kimyu0218 committed:
feat: 유효하지 않은 토큰에 대한 에러코드 추가

kimyu0218 committed:
refactor: 변경된 의존성 이름 반영 및 생성자 제거

kimyu0218 committed:
feat: delete 추가

kimyu0218 committed:
move: 상위 디렉토리 변경

refreshToken -> auth

kimyu0218 committed:
fix: delete 오류 수정

refresh 토큰 삭제 시, altKey에 postfix를 붙여 올바른 key 생성 후 삭제

kimyu0218 committed:
refactor: 테스트 케이스 추가

kimyu0218 committed:
chore: redis 추가

충돌 후 merge 과정에서 사라진 redis 다시 추가
